# Copyright (c) 2006, 2015, Oracle and/or its affiliates. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, and/or sell copies of the Software, and to permit persons
# to whom the Software is furnished to do so, provided that the above
# copyright notice(s) and this permission notice appear in all copies of
# the Software and that both the above copyright notice(s) and this
# permission notice appear in supporting documentation.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
# OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL
# INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING
# FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
# WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
# 
# Except as contained in this notice, the name of a copyright holder
# shall not be used in advertising or otherwise to promote the sale, use
# or other dealings in this Software without prior written authorization
# of the copyright holder.

# Note: BE SURE to use DRM headers from the kernel code!

diff --git a/libkms/linux.c b/libkms/linux.c
index fc4f205..5d66fc7 100644
--- libdrm-2.4.64/libkms/linux.c.~1~	2015-09-14 10:39:04.176885400 +0300
+++ libdrm-2.4.64/libkms/linux.c	2015-09-14 10:40:12.713465696 +0300
@@ -40,6 +40,7 @@
 #include <string.h>
 #include <unistd.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <sys/types.h>
 #ifdef HAVE_SYS_MKDEV_H
 #include <sys/mkdev.h>
diff --git a/xf86drm.h b/xf86drm.h
index 76eb94e..6a14120 100644
--- a/xf86drm.h
+++ b/xf86drm.h
@@ -458,6 +458,17 @@ do {	register unsigned int __old __asm("o0");		\
 #endif /* architecture */
 #endif /* __GNUC__ >= 2 */
 
+#if defined(__SUNPRO_C)
+#include <atomic.h>
+#define atomic_cmpset_int(p, c, n) ((c == atomic_cas_uint(p, c, n)) ? 1 : 0)
+#define DRM_CAS(lock,old,new,__ret)          \
+               do {                          \
+                                       unsigned int __result, __old = (old);\
+                                       __result = !atomic_cmpset_int(lock,__old,new);\
+                                       __ret = __result;          \
+                               } while(0)
+#endif
+
 #ifndef DRM_CAS
 #define DRM_CAS(lock,old,new,ret) do { ret=1; } while (0) /* FAST LOCK FAILS */
 #endif
diff --git a/xf86drmMode.c b/xf86drmMode.c
index c809c44..715f23a 100644
--- libdrm-2.4.64/xf86drmMode.c.~1~	2015-09-14 10:33:51.340507095 +0300
+++ libdrm-2.4.64/xf86drmMode.c	2015-09-14 10:36:09.881744836 +0300
@@ -834,7 +834,9 @@
 	drmClose(fd);
 	return 0;
 #endif
-	return -ENOSYS;
+
+/* for now return 0 on solaris */
+	return 0;
 }
 
 int drmModeCrtcGetGamma(int fd, uint32_t crtc_id, uint32_t size,
--- a/xf86drm.c	Fri Mar 20 07:05:48 2015
+++ b/xf86drm.c	Fri Mar 20 07:06:10 2015
@@ -1105,7 +1105,7 @@
     drm_map_t map;
 
     memclear(map);
-    map.handle = (void *)(uintptr_t)handle;
+    map.handle = (drm_handle_t)(uintptr_t)handle;
 
     if(drmIoctl(fd, DRM_IOCTL_RM_MAP, &map))
 	return -errno;
--- libdrm-2.4.64/amdgpu/amdgpu_cs.c.~1~	2015-09-14 10:42:37.518409301 +0300
+++ libdrm-2.4.64/amdgpu/amdgpu_cs.c	2015-09-14 10:42:52.951974725 +0300
@@ -25,6 +25,7 @@
 #include "config.h"
 #endif
 
+#include <alloca.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
